### PR TITLE
Fix error position for switch-statement condition and for-statement initializer

### LINKED ISSUE #7150: switch statement report incorrect error position when it fails to evaluate the condition
Steps to reproduce
------------------

```powershell
PS> cat .\about.ps1
$Test = 1;
switch ($NeverUsedBefore[0]) {
    "a" {};
}
```

Expected behavior
-----------------
Run `about.ps1` reports the correct script location.
```none
PS> .\about.ps1
Cannot index into a null array.
At F:\tmp\about.ps1:2 char:9
+ switch ($NeverUsedBefore[0]) {
+         ~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+ FullyQualifiedErrorId : NullArray
```

Actual behavior
---------------
Run `about.ps1` reports the wrong script location.
```none
PS> .\about.ps1
Cannot index into a null array.
At F:\tmp\about.ps1:1 char:1
+ $Test = 1;
+ ~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+ FullyQualifiedErrorId : NullArray
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.0.1
PSEdition                      Core
GitCommitId                    v6.0.1
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by daxian-dbw**
## PR Summary

Fix #7150

- Make switch-statement report correct error position when it fails to evaluate the condition.
- Make for-statement report correct error position when it fails to evaluate the initializer.

**Before:**
```powershell
PS> $test = 1; switch ($null[0]) { default {} }
Cannot index into a null array.
At line:1 char:1
+ $test = 1; switch ($null[0]) { default {} }
+ ~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+ FullyQualifiedErrorId : NullArray

PS> $test = 1; for ($null[0];;) {}
Cannot index into a null array.
At line:1 char:1
+ $test = 1; for ($null[0];;) {}
+ ~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+ FullyQualifiedErrorId : NullArray
```
**After:**
```powershell
PS> $test = 1; switch ($null[0]) { default {} }
Cannot index into a null array.
At line:1 char:20
+ $test = 1; switch ($null[0]) { default {} }
+                    ~~~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+ FullyQualifiedErrorId : NullArray

PS> $test = 1; for ($null[0];;) {}
Cannot index into a null array.
At line:1 char:17
+ $test = 1; for ($null[0];;) {}
+                 ~~~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+ FullyQualifiedErrorId : NullArray
```

#### For switch-statement condition, retain the same behavior as is:

**Script Content**
```powershell
PS> cat .\about2.ps1
$Test = 1..2
switch ($Test)
{
    default {'a'}
}
```

```powershell
PS> .\about2.ps1
Entering debug mode. Use h or ? for help.

Hit Line breakpoint on 'F:\tmp\about2.ps1:1'

At F:\tmp\about2.ps1:1 char:1
+ $Test = 1..2
+ ~~~~~~~~~~~~
PS>> v
At F:\tmp\about2.ps1:2 char:9
+ switch ($Test)
+         ~~~~~
PS>> v
At F:\tmp\about2.ps1:4 char:14
+     default {'a'}
+              ~~~
PS>> v
a
At F:\tmp\about2.ps1:2 char:9
+ switch ($Test)
+         ~~~~~
PS>> v
At F:\tmp\about2.ps1:4 char:14
+     default {'a'}
+              ~~~
PS>> v
a
At F:\tmp\about2.ps1:2 char:9
+ switch ($Test)
+         ~~~~~
PS>> v
```

#### For for-statement initializer, fix the incorrect breaking point position when the initializer is a simple command expression:

**Script Content**
```powershell
PS> cat .\about2.ps1
$i = 1
for ('str'.Length;
     $i -gt 0; $i--)
{
    "v"
}
```

**Before:**
```powershell
PS> .\about2.ps1
Hit Line breakpoint on 'F:\tmp\about2.ps1:1'

At F:\tmp\about2.ps1:1 char:1
+ $i = 1
+ ~~~~~~
PS>> v
At F:\tmp\about2.ps1:3 char:6
+      $i -gt 0; $i--)
+      ~~~~~~~~
PS>> v
At F:\tmp\about2.ps1:5 char:5
+     "v"
+     ~~~
PS>> v
v
At F:\tmp\about2.ps1:3 char:16
+      $i -gt 0; $i--)
+                ~~~~
PS:40>> v
At F:\tmp\about2.ps1:3 char:6
+      $i -gt 0; $i--)
+      ~~~~~~~~
PS>> v
```

**After:**
```powershell
PS> .\about2.ps1
Hit Line breakpoint on 'F:\tmp\about2.ps1:1'

At F:\tmp\about2.ps1:1 char:1
+ $i = 1
+ ~~~~~~
PS>> v
At F:\tmp\about2.ps1:2 char:6
+ for ('str'.Length;
+      ~~~~~~~~~~~~
PS>> v
At F:\tmp\about2.ps1:3 char:6
+      $i -gt 0; $i--)
+      ~~~~~~~~
PS>> v
At F:\tmp\about2.ps1:5 char:5
+     "v"
+     ~~~
PS>> v
v
At F:\tmp\about2.ps1:3 char:16
+      $i -gt 0; $i--)
+                ~~~~
PS>> v
At F:\tmp\about2.ps1:3 char:6
+      $i -gt 0; $i--)
+      ~~~~~~~~
PS>> v
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/Compiler.cs [hunk 1]
```diff
@@ -903,17 +906,31 @@ internal static Expression IsStrictMode(int version, Expression executionContext
                                    ExpressionCache.Constant(version));
         }
 
-        internal Expression UpdatePosition(Ast ast)
+        private int AddSequencePoint(IScriptExtent extent)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/Compiler.cs]: Note: this method is added to make sure that one extent can only be added to the sequence list once.

**Comment by daxian-dbw** [comment]
@BrucePay Can you please review this PR when you have time? Thanks!

File: test/powershell/Language/Scripting/ErrorPosition.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        if (-not ("SwitchTest.Test" -as [type])) {
+            Add-Type -TypeDefinition $code
+        }
+
+        try { & $testFile > $null } catch { $errorRecord = $_ }
+        $errorRecord | Should -Not -BeNullOrEmpty
+        $errorRecord.ScriptStackTrace | Should -Match "SwitchError2.ps1: line 3"
+    }
+
+    It "for statement initializer evaluation failure should report correct error position" {
```
Comment (SteveL-MSFT) [inline test/powershell/Language/Scripting/ErrorPosition.Tests.ps1]: There should also be a test for while/do loops

File: test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1 [hunk 3]
```diff
@@ -8,32 +9,83 @@ $script1 = @'
...
+$test = 1..2
+switch ($test)
+{
+    default {}
+}
+'@
+
+$script3 = @'
+$test = 2
+for ("string".Length;
```
Comment (SteveL-MSFT) [inline test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1]: In the product code, there's a special check for while/do loops, but I don't see a test here for that?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Review by SteveL-MSFT — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]
LGTM

File: src/System.Management.Automation/engine/parser/Compiler.cs [hunk 4]
```diff
@@ -4240,10 +4258,26 @@ private Expression GenerateDoLoop(LoopStatementAst loopStatement)
...
             if (generatingForeach)
             {
+                // For foreach statement, we want the debugger to stop at 'stmt.Condition' before evaluating it.
+                // The debugger will stop at 'stmt.Condition' only once. The following enumeration will stop at
+                // the foreach variable.
                 exprs.Add(UpdatePosition(stmt.Condition));
             }
+            else
+            {
+                // For switch statement, we don't want the debugger to stop at 'stmt.Condition' before evaluating it.
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: The `checkBreakpoints: false` when calling `UpdatePositionExpr` means not checking breaking point for this sequence point update.

File: test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1 [hunk 5]
```diff
@@ -1,39 +1,296 @@
...
 "a" 2> $null | ForEach-Object { $_ }
 'bb' > $null
 'bb'.ToSTring() > $null
 'bbb'
 '@
-$script2 = @'
-"line 1"
-"line 2"
-"line 3"
+            $path = Setup -PassThru -File BasicTest.ps1 -Content $script
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1]: That sounds good. Can you please show an example of how to check for debugger stop position?

[the remark this answers is not in the document]
Comment (PaulHigin) [reply test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1]: You can create a type (Add-Type) that has a method to handle the runspace.debugger.DebuggerStop event and then check the event argument InvocationInfo extent (I think).  Contact me offline for more info.

[the remark this answers is not in the document]
Comment (daxian-dbw) [reply test/powershell/Language/Scripting/Debugging/Debugging.Tests.ps1]: OK. I will play with it, and add more tests in a separate PR.

**Pull request merged**